storage: TestStoreRangeRemoveDead failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/9f191cdad64747f7f19ec4990e559cbb19e6d372

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestStoreRangeRemoveDead PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=928632&tab=buildLog

```
I180927 06:07:20.113115 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:21.126234 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:22.138801 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:23.151225 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:24.157603 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:25.165726 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:26.186466 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:27.188758 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:28.191561 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
I180927 06:07:29.194531 68674 storage/client_raft_test.go:2906  SucceedsSoon: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
W180927 06:07:30.200435 68763 internal/client/txn.go:532  [hb] failure aborting transaction: node unavailable; try another peer; abort caused by: result is ambiguous (server shutdown)
I180927 06:07:30.204310 68763 storage/node_liveness.go:830  [hb] retrying liveness update after storage.errRetryLiveness: result is ambiguous (server shutdown)
W180927 06:07:30.220042 68763 internal/client/txn.go:532  [hb] failure aborting transaction: node unavailable; try another peer; abort caused by: node unavailable; try another peer
W180927 06:07:30.220476 68763 storage/node_liveness.go:494  [hb] failed node liveness heartbeat: node unavailable; try another peer
W180927 06:07:30.222988 69459 storage/store.go:1657  [s6,r1/3:/M{in-ax}] could not gossip first range descriptor: node unavailable; try another peer
W180927 06:07:30.231595 69479 storage/raft_transport.go:281  unable to accept Raft message from (n4,s4):?: no handler registered for (n1,s1):?
W180927 06:07:30.233204 69373 storage/store.go:3820  [s4] raft error: node 1 claims to not contain store 1 for replica (n1,s1):?: store 1 was not found
W180927 06:07:30.233573 69371 storage/raft_transport.go:583  while processing outgoing Raft queue to node 1: store 1 was not found:
W180927 06:07:30.289119 69506 storage/raft_transport.go:583  while processing outgoing Raft queue to node 5: rpc error: code = Canceled desc = grpc: the client connection is closing:
W180927 06:07:30.294849 69480 storage/raft_transport.go:583  while processing outgoing Raft queue to node 6: rpc error: code = Canceled desc = grpc: the client connection is closing:
--- FAIL: TestStoreRangeRemoveDead (46.88s)
	soon.go:49: condition failed to evaluate within 45s: expected 5 replicas; have [(n1,s1):1 (n4,s4):2 (n6,s6):3 (n2,s2):4 (n3,s3):5 (n5,s5):6]
		goroutine 68674 [running]:
		runtime/debug.Stack(0xa7a358200, 0xc420b1e5d0, 0x3bd9f60)
			/usr/local/go/src/runtime/debug/stack.go:24 +0xb5
		github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x3c35e00, 0xc4215163c0, 0xc4219b01e0)
			/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:50 +0x172
		github.com/cockroachdb/cockroach/pkg/storage_test.TestStoreRangeRemoveDead(0xc4215163c0)
			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_raft_test.go:2906 +0x686
		testing.tRunner(0xc4215163c0, 0x34caf48)
			/usr/local/go/src/testing/testing.go:777 +0x16e
		created by testing.(*T).Run
			/usr/local/go/src/testing/testing.go:824 +0x565

```

SHA: https://github.com/cockroachdb/cockroach/commits/e768822f584e0150b04323183e3b0360b9c1ad8b

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestStoreRangeRemoveDead PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=929112&tab=buildLog

```
I180927 06:58:57.357394 69557 storage/replica_raftstorage.go:803  [s3,r1/?:{-}] applying preemptive snapshot at index 35 (id=4f156ca0, encoded size=15882, 1 rocksdb batches, 25 log entries)
I180927 06:58:57.363871 69557 storage/replica_raftstorage.go:809  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 6ms [clear=0ms batch=0ms entries=5ms commit=0ms]
I180927 06:58:57.368855 68860 storage/replica_command.go:812  [replicate,s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):6): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n5,s5):4, next=6, gen=0]
I180927 06:58:57.403018 68860 storage/replica.go:3761  [replicate,s1,r1/1:/M{in-ax},txn=84f95f35] proposing ADD_REPLICA((n3,s3):6): updated=[(n1,s1):1 (n2,s2):2 (n5,s5):4 (n3,s3):6] next=7
I180927 06:58:57.422387 68860 storage/store_snapshot.go:615  [replicate,s1,r1/1:/M{in-ax}] sending preemptive snapshot d58ef5aa at applied index 37
I180927 06:58:57.424658 68860 storage/store_snapshot.go:657  [replicate,s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 77, log entries: 27, rate-limit: 8.0 MiB/sec, 5ms
I180927 06:58:57.426742 69259 storage/replica_raftstorage.go:803  [s4,r1/?:{-}] applying preemptive snapshot at index 37 (id=d58ef5aa, encoded size=16998, 1 rocksdb batches, 27 log entries)
I180927 06:58:57.436215 69259 storage/replica_raftstorage.go:809  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 9ms [clear=0ms batch=0ms entries=8ms commit=0ms]
I180927 06:58:57.440264 68860 storage/replica_command.go:812  [replicate,s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):7): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n5,s5):4, (n3,s3):6, next=7, gen=0]
I180927 06:58:57.466663 68860 storage/replica.go:3761  [replicate,s1,r1/1:/M{in-ax},txn=497bae96] proposing ADD_REPLICA((n4,s4):7): updated=[(n1,s1):1 (n2,s2):2 (n5,s5):4 (n3,s3):6 (n4,s4):7] next=8
I180927 06:58:57.507324 68860 storage/replica_command.go:812  [replicate,s1,r1/1:/M{in-ax}] change replicas (REMOVE_REPLICA (n5,s5):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n5,s5):4, (n3,s3):6, (n4,s4):7, next=8, gen=0]
I180927 06:58:57.543954 68860 storage/replica.go:3761  [replicate,s1,r1/1:/M{in-ax},txn=d16b7f1e] proposing REMOVE_REPLICA((n5,s5):4): updated=[(n1,s1):1 (n2,s2):2 (n4,s4):7 (n3,s3):6] next=8
I180927 06:58:57.555081 69128 storage/store.go:3794  [s5,r1/4:/M{in-ax}] added to replica GC queue (peer suggestion)
I180927 06:58:57.558121 69558 storage/store.go:2735  [replicaGC,s5,r1/4:/M{in-ax}] removing replica r1/4
I180927 06:58:57.559945 69558 storage/replica.go:871  [replicaGC,s5,r1/4:/M{in-ax}] removed 50 (45+5) keys in 1ms [clear=1ms commit=0ms]
I180927 06:58:58.499650 68875 gossip/gossip.go:940  [n1] gossip connectivity
  n4 [sentinel];
  n2 -> n1; n3 -> n1; n4 -> n1; n5 -> n4;
I180927 06:58:58.554265 69117 gossip/gossip.go:940  [n2] gossip connectivity
  n4 [sentinel];
  n2 -> n1; n3 -> n1; n4 -> n1; n5 -> n4;
I180927 06:58:58.656793 69240 gossip/gossip.go:940  [n3] gossip connectivity
  n4 [sentinel];
  n2 -> n1; n3 -> n1; n4 -> n1; n5 -> n4;
I180927 06:58:58.777267 69379 gossip/gossip.go:940  [n4] gossip connectivity
  n4 [sentinel];
  n2 -> n1; n3 -> n1; n4 -> n1; n5 -> n4;
I180927 06:58:58.834368 69483 gossip/gossip.go:940  [n5] gossip connectivity
  n4 [sentinel];
  n2 -> n1; n3 -> n1; n4 -> n1; n5 -> n4;
I180927 06:58:58.931960 69240 gossip/gossip.go:1533  [n3] node has connected to cluster via gossip
I180927 06:58:58.963657 69379 gossip/gossip.go:1533  [n4] node has connected to cluster via gossip
I180927 06:58:59.004683 69117 gossip/gossip.go:1533  [n2] node has connected to cluster via gossip
W180927 06:59:02.118628 69546 storage/raft_transport.go:583  while processing outgoing Raft queue to node 3: rpc error: code = Canceled desc = grpc: the client connection is closing:
W180927 06:59:02.119416 69117 gossip/gossip.go:1519  [n2] no incoming or outgoing connections
--- FAIL: TestStoreRangeRemoveDead (5.69s)
	client_raft_test.go:2897: Failed to remove the dead replica within 5s

```

--- a/pkg/storage/client_raft_test.go
+++ b/pkg/storage/client_raft_test.go
@@ -2858,8 +2858,16 @@ func TestStoreRangeRemoveDead(t *testing.T) {
 	zone := config.DefaultSystemZoneConfig()
 	mtc.Start(t, int(zone.NumReplicas+1))
 
+	var nonDeadStores []*storage.Store
+	for i, s := range mtc.stores {
+		if i == 1 {
+			// Skip the dead store.
+			continue
+		}
+		nonDeadStores = append(nonDeadStores, s)
+	}
+
 	// Create a goroutine to gossip store capacity info periodically.
-	var deadStoreID int32 = -1
 	go func() {
 		tickerDur := storage.TestTimeUntilStoreDead / 2
 		ticker := time.NewTicker(tickerDur)
@@ -2871,18 +2879,14 @@ func TestStoreRangeRemoveDead(t *testing.T) {
 				mtc.manualClock.Increment(int64(tickerDur))
 
 				// Keep gossiping the stores, excepting the dead store.
-				for _, s := range mtc.stores {
-					if s.Ident.StoreID != roachpb.StoreID(atomic.LoadInt32(&deadStoreID)) {
-						if err := s.GossipStore(context.Background(), false /* useCached */); err != nil {
-							panic(err)
-						}
+				for _, s := range nonDeadStores {
+					if err := s.GossipStore(context.Background(), false /* useCached */); err != nil {
+						panic(err)
 					}
 				}
 				// Force the repair queues on all alive stores to run.
-				for _, s := range mtc.stores {
-					if s.Ident.StoreID != roachpb.StoreID(atomic.LoadInt32(&deadStoreID)) {
-						s.ForceReplicationScanAndProcess()
-					}
+				for _, s := range nonDeadStores {
+					s.ForceReplicationScanAndProcess()
 				}
 
 			case <-mtc.stoppers[0].ShouldStop():
@@ -2900,16 +2904,21 @@ func TestStoreRangeRemoveDead(t *testing.T) {
 		return errors.Errorf("expected %d replicas; have %+v", zone.NumReplicas, replicas)
 	})
 
-	// Set the dead store and wait for the replica on the dead node to
-	// be removed and the replacement added. Use a store other than the leader.
-	atomic.StoreInt32(&deadStoreID, int32(getRangeMetadata(roachpb.RKeyMin, mtc, t).Replicas[1].StoreID))
+	// Stop a store which will be rebalanced away from. We can't use the very first
+	// one since getRangeMetadata is hard-coded to query that one, so we'll use the
+	// one after that.
+	deadStoreID := mtc.stores[1].StoreID() // 2
+	mtc.stopStore(1)
+	// The mtc asks us to restart this store before stopping the mtc.
+	defer mtc.restartStoreWithoutHeartbeat(1)
+
 	testutils.SucceedsSoon(t, func() error {
 		replicas := getRangeMetadata(roachpb.RKeyMin, mtc, t).Replicas
 		if len(replicas) != int(zone.NumReplicas) {
 			return errors.Errorf("expected %d replicas; have %+v", zone.NumReplicas, replicas)
 		}
 		for _, r := range replicas {
-			if r.StoreID == roachpb.StoreID(atomic.LoadInt32(&deadStoreID)) {
+			if r.StoreID == deadStoreID {
 				return errors.Errorf("expected store %d to be replaced; have %+v", r.StoreID, replicas)
 			}
 		}

